### DISCUSSION TITLE
how to filter out source files without declarations during emission?

### DISCUSSION THREAD

**Discussion by boan-anbo** [microsoft/typespec#4702] [Q&A] [answered]
Hello, my use case is to generate file by namespace.

I `createSourceFile` in each `namespaceContext` call, but that means generating files for all namespaces, empty or not.

I only want to write non-empty source files.

I could access the asset emitter after the emission but only to get information about source files and their declarations in globalscope, I don't know what I can do to filter these out before calling `writeOutput()`.

Or perhaps I was not following the correct control flow in creating source file in the first place?

Other than the empty files for empty namespaces, the generated files are OK.

Thank you.

**Comment by timotheeguerin** [comment, marked as answer]
I asume you are talking about the emitter framework, if you registered your declarations on the right scope you should be able to get all declarations added to a file and decide if you want to emit it or not.

https://github.com/microsoft/typespec/blob/ff1725c4a833a81b54626f0d6c847edab9e8dd74/packages/json-schema/src/json-schema-emitter.ts#L720

Side note: emitter framework in its current state as a lot of issues and is not ideal, we are working on a completely different approach to help emitter authors

**Reply by boan-anbo** [reply]
Got it. Thanks so much and look forward to the alternative approach!
